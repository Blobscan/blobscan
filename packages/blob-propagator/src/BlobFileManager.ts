import fs from "fs";
import path from "path";

import type { Blob } from "./types";

export type BlobFileManagerConfig = {
  basePath: string;
  folderName: string;
};

export class BlobFileManager {
  #blobFolderPath: string;

  constructor({
    basePath = __dirname,
    folderName = "eip4844-blobs",
  }: Partial<BlobFileManagerConfig> = {}) {
    this.#blobFolderPath = path.join(basePath, folderName);

    if (!fs.existsSync(this.#blobFolderPath)) {
      fs.mkdirSync(this.#blobFolderPath);
    }
  }

  checkFileExists(versionedHash: string) {
    const path = this.#buildFilePath(versionedHash);

    return fs.promises
      .access(path, fs.constants.F_OK)
      .then(() => true)
      .catch(() => false);
  }

  async createFile({ data, versionedHash }: Blob) {
    try {
      if (await this.checkFileExists(versionedHash)) {
        return Promise.resolve();
      }

      const blobfilePath = this.#buildFilePath(versionedHash);

      return fs.promises.writeFile(blobfilePath, data, { encoding: "utf-8" });
    } catch (err) {
      throw new Error(
        `Couldn't store blob ${versionedHash} data on filesystem: ${err}`
      );
    }
  }

  async readFile(versionedHash: string) {
    const blobFilePath = this.#buildFilePath(versionedHash);

    try {
      const blobData = await fs.promises.readFile(blobFilePath, "utf-8");

      return blobData;
    } catch (error) {
      throw new Error(
        `Couldn't read blob ${versionedHash} data file: file is missing`
      );
    }
  }

  async removeFile(versionedHash: string) {
    try {
      await fs.promises.unlink(this.#buildFilePath(versionedHash));
    } catch (err) {
      throw new Error(
        `Couldn't remove blob ${versionedHash} data file: ${err}`
      );
    }
  }

  async removeFolder() {
    return fs.promises.rm(this.#blobFolderPath, {
      recursive: true,
    });
  }

  #buildFilePath(versionedHash: string) {
    return path.join(this.#blobFolderPath, `${versionedHash}.txt`);
  }
}
