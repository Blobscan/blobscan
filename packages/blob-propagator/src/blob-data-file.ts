import fs from "fs";

import type { Blob } from "./types";

export function fileExists(path: string) {
  return fs.promises
    .access(path, fs.constants.F_OK)
    .then(() => true)
    .catch(() => false);
}

export function buildBlobDataFilePath(versionedHash: string) {
  return `${versionedHash}.txt`;
}

export async function createBlobDataFile({ data, versionedHash }: Blob) {
  const blobfilePath = buildBlobDataFilePath(versionedHash);

  try {
    if (await fileExists(blobfilePath)) {
      return Promise.resolve();
    }

    return fs.promises.writeFile(blobfilePath, data, { encoding: "utf-8" });
  } catch (err) {
    throw new Error(
      `Couldn't store blob ${versionedHash} data on filesystem: ${err}`
    );
  }
}

export async function readBlobDataFile(versionedHash: string) {
  const blobFilePath = buildBlobDataFilePath(versionedHash);

  try {
    const blobData = await fs.promises.readFile(blobFilePath, "utf-8");

    return blobData;
  } catch (error) {
    throw new Error(
      `Couldn't read blob ${versionedHash} data file: file is missing`
    );
  }
}

export async function removeBlobDataFile(versionedHash: string) {
  try {
    await fs.promises.unlink(buildBlobDataFilePath(versionedHash));
  } catch (err) {
    throw new Error(`Couldn't remove blob ${versionedHash} data file: ${err}`);
  }
}
