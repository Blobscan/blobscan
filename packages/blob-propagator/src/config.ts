import type { QueueOptions, WorkerOptions } from "bullmq";

import type { BlobStorage } from "@blobscan/db";

import { env } from "./env";

export const STORAGE_QUEUES: Record<BlobStorage, string> = {
  GOOGLE: "gcs-queue",
  SWARM: "swarm-queue",
  POSTGRES: "postgres-queue",
};

export const FINALIZER_QUEUE = "finalizer-queue";

export const DEFAULT_OPTIONS: QueueOptions = {
  defaultJobOptions: {
    backoff: {
      type: "exponential",
    },
    attempts: 3,
    removeOnComplete: true,
  },
};

export const DEFAULT_WORKER_OPTIONS: WorkerOptions = {
  useWorkerThreads: true,
  connection: {
    host: env.REDIS_QUEUE_HOST,
    port: env.REDIS_QUEUE_PORT,
    password: env.REDIS_QUEUE_PASSWORD,
    username: env.REDIS_QUEUE_USERNAME,
  },
};
