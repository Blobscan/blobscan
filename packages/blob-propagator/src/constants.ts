import type { JobsOptions, WorkerOptions } from "bullmq";

import { $Enums } from "@blobscan/db";

import { env } from "./env";

export const STORAGE_WORKER_NAMES = Object.values($Enums.BlobStorage).reduce<
  Record<$Enums.BlobStorage, string>
>(
  (names, storage) => ({
    ...names,
    [storage]: `${storage.toLowerCase()}-worker`,
  }),
  {} as Record<$Enums.BlobStorage, string>
);

export const FINALIZER_WORKER_NAME = "finalizer-worker";

export const DEFAULT_JOB_OPTIONS: Omit<JobsOptions, "repeat"> = {
  attempts: 3,
  backoff: {
    type: "exponential",
    delay: 1000,
  },
};

export const DEFAULT_WORKER_OPTIONS: WorkerOptions = {
  autorun: true,
  useWorkerThreads: false,
  removeOnComplete: { age: env.BLOB_PROPAGATOR_COMPLETED_JOBS_AGE },
  removeOnFail: { age: env.BLOB_PROPAGATOR_FAILED_JOBS_AGE },
};
