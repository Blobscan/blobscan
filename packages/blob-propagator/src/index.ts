export type { EnvVars as Environment } from "./env";

export { env } from "./env";
export { BlobPropagator } from "./BlobPropagator";
export { blobPropagator } from "./blob-propagator";
export * from "./types";
export {
  createBlobPropagationFlowJob,
  createBlobStorageJob,
  FINALIZER_WORKER_NAME,
  STORAGE_WORKER_NAMES,
  buildJobId,
} from "./utils";
