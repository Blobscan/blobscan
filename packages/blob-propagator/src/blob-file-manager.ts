import os from "os";

import { BlobFileManager } from "./BlobFileManager";
import { env } from "./env";

const basePath = !env.TEST ? os.tmpdir() : undefined;

export const blobFileManager = new BlobFileManager({
  basePath,
  folderName: "blobscan-blobs",
});
