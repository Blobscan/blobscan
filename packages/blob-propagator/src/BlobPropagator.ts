/* eslint-disable @typescript-eslint/no-misused-promises */
import { FlowProducer, Queue, Worker } from "bullmq";
import type { ConnectionOptions, FlowJob, WorkerOptions } from "bullmq";

import type { $Enums } from "@blobscan/db";
import type { BlobStorage } from "@blobscan/db";
import { logger } from "@blobscan/logger";

import { blobFileManager } from "./blob-file-manager";
import type { Blob, BlobPropagationJobData } from "./types";
import {
  FINALIZER_WORKER_NAME,
  STORAGE_WORKER_NAMES,
  createBlobPropagationFlowJob,
} from "./utils";
import {
  fileSystemProcessor,
  finalizerProcessor,
  gcsProcessor,
  postgresProcessor,
  swarmProcessor,
} from "./worker-processors";

const STORAGE_WORKER_PROCESSORS = {
  GOOGLE: gcsProcessor,
  SWARM: swarmProcessor,
  POSTGRES: postgresProcessor,
  FILE_SYSTEM: fileSystemProcessor,
};

export type BlobPropagatorOptions = Partial<{
  workerOptions: WorkerOptions;
}>;

const DEFAULT_WORKER_OPTIONS: WorkerOptions = {
  autorun: true,
  useWorkerThreads: false,
  removeOnComplete: { count: 1000 },
};

export class BlobPropagator {
  protected blobPropagationFlowProducer: FlowProducer;
  protected finalizerWorker: Worker;
  protected storageWorkers: Record<
    $Enums.BlobStorage,
    Worker<BlobPropagationJobData>
  >;

  constructor(
    storages: $Enums.BlobStorage[],
    { workerOptions }: BlobPropagatorOptions = {}
  ) {
    if (!storages.length) {
      throw new Error(
        "Couldn't instantiate blob propagator: no storages given"
      );
    }

    this.storageWorkers = this.#createBlobStorageWorkers(
      storages,
      workerOptions
    );
    this.finalizerWorker = this.#createFinalizerWorker(workerOptions);
    this.blobPropagationFlowProducer = this.#createBlobPropagationFlowProducer(
      workerOptions?.connection
    );
  }

  async empty({ force }: { force: boolean } = { force: false }) {
    const workers = this.#getWorkers();
    const queues = await Promise.all(
      workers.map(
        async (w) => new Queue(w.name, { connection: await w.client })
      )
    );

    let emptyPromise = Promise.all([
      ...queues.map((q) => q.obliterate({ force })),
      blobFileManager.removeFolder(),
    ]);

    queues.forEach((q) => {
      emptyPromise = emptyPromise.finally(async () => {
        await q.close();
      });
    });

    await emptyPromise;
  }

  close() {
    let teardownPromise: Promise<void> = Promise.resolve();

    Object.values(this.storageWorkers).forEach((w) => {
      teardownPromise = teardownPromise.finally(async () => {
        await w.close();
      });
    });

    return teardownPromise
      .finally(async () => {
        await this.finalizerWorker.close();
      })
      .finally(async () => {
        const redisClient = await this.blobPropagationFlowProducer.client;

        await redisClient.quit();
      })
      .finally(async () => {
        await this.blobPropagationFlowProducer.close();
      });
  }

  async propagateBlob(blob: Blob) {
    await blobFileManager.createFile(blob);

    const blobFlowJob = await this.#createBlobPropagationFlowJob({
      versionedHash: blob.versionedHash,
    });

    await this.blobPropagationFlowProducer.add(blobFlowJob);
  }

  async propagateBlobs(blobs: Blob[]) {
    const uniqueBlobs = Array.from(
      new Set(blobs.map((b) => b.versionedHash))
    ).map((versionedHash) => {
      // @typescript-eslint/no-non-null-assertion
      const blob = blobs.find((b) => b.versionedHash === versionedHash)!;

      return blob;
    });

    await Promise.all(
      uniqueBlobs.map((blob) => blobFileManager.createFile(blob))
    );

    const blobPropagationFlowJobs = uniqueBlobs.map(({ versionedHash }) =>
      this.#createBlobPropagationFlowJob({ versionedHash })
    );

    await this.blobPropagationFlowProducer.addBulk(blobPropagationFlowJobs);
  }

  #getWorkers() {
    return [...Object.values(this.storageWorkers), this.finalizerWorker];
  }

  #createBlobPropagationFlowProducer(connection?: ConnectionOptions) {
    /*
     * Instantiating a new `FlowProducer` appears to create two separate `RedisConnection` instances.
     * This leads to an issue where one instance remains active, or "dangling", after the `FlowProducer` has been closed.
     * To prevent this, we now initialize a single `IORedis` instance in advance and use it when creating the `FlowProducer`.
     * This way, both created connections reference the same `IORedis` instance and can be closed properly.
     *
     * See: https://github.com/taskforcesh/bullmq/blob/d7cf6ea60830b69b636648238a51e5f981616d02/src/classes/flow-producer.ts#L111
     */
    const blobPropagationFlowProducer = new FlowProducer({
      connection,
    });

    blobPropagationFlowProducer.on("error", (err) => {
      logger.error(`Blob propagation flow producer error: ${err}`);
    });

    return blobPropagationFlowProducer;
  }

  #createFinalizerWorker(opts: WorkerOptions = {}) {
    const finalizerWorker = new Worker(
      FINALIZER_WORKER_NAME,
      finalizerProcessor,
      {
        ...DEFAULT_WORKER_OPTIONS,
        ...opts,
      }
    );

    finalizerWorker.on("completed", (job) => {
      logger.debug(`Job ${job.id} completed`);
    });

    finalizerWorker.on("failed", (job, err) => {
      logger.error(`Job ${job?.id} failed: ${err}`);
    });

    return finalizerWorker;
  }

  #createBlobStorageWorkers(
    storages: $Enums.BlobStorage[],
    opts: WorkerOptions = {}
  ) {
    return storages.reduce<Record<BlobStorage, Worker<BlobPropagationJobData>>>(
      (workers, storageName) => {
        const workerName = STORAGE_WORKER_NAMES[storageName];
        const storageWorker = new Worker<BlobPropagationJobData>(
          workerName,
          STORAGE_WORKER_PROCESSORS[storageName],
          {
            ...DEFAULT_WORKER_OPTIONS,
            ...opts,
          }
        );

        storageWorker.on("completed", (job) => {
          logger.debug(`Job ${job.id} completed by ${workerName}`);
        });

        storageWorker.on("failed", (job, err) => {
          logger.error(`Job ${job?.id} failed: ${err} (worker: ${workerName})`);
        });

        workers[storageName] = storageWorker;

        return workers;
      },
      {} as Record<BlobStorage, Worker<BlobPropagationJobData>>
    );
  }

  #createBlobPropagationFlowJob(data: BlobPropagationJobData): FlowJob {
    const versionedHash = data.versionedHash;

    const storageWorkerNames = Object.values(this.storageWorkers).map(
      (storageWorker) => storageWorker.name
    );

    return createBlobPropagationFlowJob(
      this.finalizerWorker.name,
      storageWorkerNames,
      versionedHash
    );
  }
}
