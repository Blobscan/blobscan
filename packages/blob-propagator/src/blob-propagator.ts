import IORedis from "ioredis";

import { BlobStorage } from "@blobscan/db";

import { BlobPropagator } from "./BlobPropagator";
import { env } from "./env";

function createBlobPropagator() {
  const availableStorages: BlobStorage[] = [];

  if (env.GOOGLE_STORAGE_ENABLED) {
    availableStorages.push(BlobStorage.GOOGLE);
  }

  if (env.POSTGRES_STORAGE_ENABLED) {
    availableStorages.push(BlobStorage.POSTGRES);
  }

  if (env.SWARM_STORAGE_ENABLED) {
    availableStorages.push(BlobStorage.SWARM);
  }

  const connection = new IORedis(env.REDIS_URI, { maxRetriesPerRequest: null });

  return new BlobPropagator(availableStorages, {
    workerOptions: {
      connection,
    },
  });
}
const blobPropagator =
  env.BLOB_PROPAGATOR_ENABLED === true ? createBlobPropagator() : undefined;

export { blobPropagator, createBlobPropagator };
