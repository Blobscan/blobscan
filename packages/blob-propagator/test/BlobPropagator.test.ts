import type { ConnectionOptions } from "bullmq";
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";

import { BlobStorage } from "@blobscan/db";

import { BlobPropagator } from "../src/BlobPropagator";
import { blobFileManager } from "../src/blob-file-manager";
import type { Blob } from "../src/types";

export class MockedBlobPropagator extends BlobPropagator {
  getFinalizerWorker() {
    return this.finalizerWorker;
  }

  getStorageWorkers() {
    return this.storageWorkers;
  }

  getBlobPropagationFlowProducer() {
    return this.blobPropagationFlowProducer;
  }
}

describe("BlobPropagator", () => {
  const blobStorages = Object.values(BlobStorage);
  const connection: ConnectionOptions = {
    host: "localhost",
    port: 6379,
  };
  const blob: Blob = {
    versionedHash: "blobVersionedHash",
    data: "0x1234abcdeff123456789ab34223a4b2c2e",
  };

  let blobPropagator: MockedBlobPropagator;

  beforeEach(() => {
    blobPropagator = new MockedBlobPropagator(blobStorages, {
      workerOptions: {
        connection,
      },
    });
  });

  afterEach(async () => {
    await blobPropagator.close();
  });

  it("should throw an error when trying to instantiate it with no storages", () => {
    expect(
      () => new MockedBlobPropagator([])
    ).toThrowErrorMatchingInlineSnapshot(
      '"Couldn\'t instantiate blob propagator: no storages given"'
    );
  });

  it("should close correctly", async () => {
    const testBlobPropagator = new MockedBlobPropagator(blobStorages, {
      workerOptions: {
        connection,
      },
    });
    // Flow producer doesn't have a running state, so we need to spy on its close method
    const flowProducer = testBlobPropagator.getBlobPropagationFlowProducer();
    const flowProducerCloseSpy = vi.spyOn(flowProducer, "close");

    await testBlobPropagator.close();

    const finalizerClosed = !testBlobPropagator
      .getFinalizerWorker()
      .isRunning();
    const storageWorkersClosed = Object.values(
      testBlobPropagator.getStorageWorkers()
    ).every((worker) => !worker.isRunning());

    expect(finalizerClosed, "Finalizer worker still running").toBe(true);
    expect(storageWorkersClosed, "Storage workers still running").toBe(true);
    expect(
      flowProducerCloseSpy,
      "Flow producer still running"
    ).toHaveBeenCalled();
  });

  describe("when propagating a single blob", () => {
    afterEach(async () => {
      blobFileManager.removeFile(blob.versionedHash);
    });

    it("should store blob data on disk", async () => {
      await blobPropagator.propagateBlob(blob);

      const fileExists = await blobFileManager.checkFileExists(
        blob.versionedHash
      );

      expect(fileExists).toBe(true);
    });

    it("should create a blob propagation flow job correctly", async () => {
      const propagatorFlowProducer =
        blobPropagator.getBlobPropagationFlowProducer();

      const spy = vi.spyOn(propagatorFlowProducer, "add");

      await blobPropagator.propagateBlob(blob);

      const createdFlowJob = spy.mock.calls[0]?.[0];

      expect(createdFlowJob).toMatchInlineSnapshot(`
        {
          "children": [
            {
              "data": {
                "versionedHash": "blobVersionedHash",
              },
              "name": "storeBlob:file_system-worker-blobVersionedHash",
              "opts": {
                "attempts": 3,
                "backoff": {
                  "delay": 1000,
                  "type": "exponential",
                },
                "jobId": "file_system-worker-blobVersionedHash",
              },
              "queueName": "file_system-worker",
            },
            {
              "data": {
                "versionedHash": "blobVersionedHash",
              },
              "name": "storeBlob:google-worker-blobVersionedHash",
              "opts": {
                "attempts": 3,
                "backoff": {
                  "delay": 1000,
                  "type": "exponential",
                },
                "jobId": "google-worker-blobVersionedHash",
              },
              "queueName": "google-worker",
            },
            {
              "data": {
                "versionedHash": "blobVersionedHash",
              },
              "name": "storeBlob:postgres-worker-blobVersionedHash",
              "opts": {
                "attempts": 3,
                "backoff": {
                  "delay": 1000,
                  "type": "exponential",
                },
                "jobId": "postgres-worker-blobVersionedHash",
              },
              "queueName": "postgres-worker",
            },
            {
              "data": {
                "versionedHash": "blobVersionedHash",
              },
              "name": "storeBlob:swarm-worker-blobVersionedHash",
              "opts": {
                "attempts": 3,
                "backoff": {
                  "delay": 1000,
                  "type": "exponential",
                },
                "jobId": "swarm-worker-blobVersionedHash",
              },
              "queueName": "swarm-worker",
            },
          ],
          "data": {
            "versionedHash": "blobVersionedHash",
          },
          "name": "propagateBlob:finalizer-worker-blobVersionedHash",
          "opts": {
            "attempts": 3,
            "backoff": {
              "delay": 1000,
              "type": "exponential",
            },
            "jobId": "finalizer-worker-blobVersionedHash",
          },
          "queueName": "finalizer-worker",
        }
      `);
    });
  });

  describe("when propagating multiple blobs", () => {
    const blobs: Blob[] = [
      {
        versionedHash: "blobVersionedHash1",
        data: "0x1234abcdeff123456789ab34223a4b2c2e",
      },
      {
        versionedHash: "blobVersionedHash2",
        data: "0x2467234ab65cff34faa23",
      },
    ];

    afterEach(async () => {
      await Promise.all(
        blobs.map(async (b) => {
          if (await blobFileManager.checkFileExists(b.versionedHash)) {
            blobFileManager.removeFile(b.versionedHash);
          }
        })
      );
    });

    it("should store blob data on disk", async () => {
      await blobPropagator.propagateBlobs(blobs);

      const filesExists = (
        await Promise.all(
          blobs.map((b) => blobFileManager.checkFileExists(b.versionedHash))
        )
      ).every((exists) => exists);

      expect(filesExists).toBe(true);
    });

    it("should create flow jobs correctly", async () => {
      const blobPropagationFlowProducer =
        blobPropagator.getBlobPropagationFlowProducer();

      const spy = vi.spyOn(blobPropagationFlowProducer, "addBulk");

      await blobPropagator.propagateBlobs(blobs);

      const createdFlowJobs = spy.mock.calls[0]?.[0];

      expect(createdFlowJobs).toMatchSnapshot();
    });
  });
});
