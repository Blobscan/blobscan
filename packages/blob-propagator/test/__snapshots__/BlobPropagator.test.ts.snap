// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BlobPropagator > when propagating multiple blobs > should create flow jobs correctly 1`] = `
[
  {
    "children": [
      {
        "data": {
          "versionedHash": "blobVersionedHash1",
        },
        "name": "storeBlob:file_system-worker-blobVersionedHash1",
        "opts": {
          "attempts": 3,
          "backoff": {
            "delay": 1000,
            "type": "exponential",
          },
          "jobId": "file_system-worker-blobVersionedHash1",
        },
        "queueName": "file_system-worker",
      },
      {
        "data": {
          "versionedHash": "blobVersionedHash1",
        },
        "name": "storeBlob:google-worker-blobVersionedHash1",
        "opts": {
          "attempts": 3,
          "backoff": {
            "delay": 1000,
            "type": "exponential",
          },
          "jobId": "google-worker-blobVersionedHash1",
        },
        "queueName": "google-worker",
      },
      {
        "data": {
          "versionedHash": "blobVersionedHash1",
        },
        "name": "storeBlob:postgres-worker-blobVersionedHash1",
        "opts": {
          "attempts": 3,
          "backoff": {
            "delay": 1000,
            "type": "exponential",
          },
          "jobId": "postgres-worker-blobVersionedHash1",
        },
        "queueName": "postgres-worker",
      },
      {
        "data": {
          "versionedHash": "blobVersionedHash1",
        },
        "name": "storeBlob:swarm-worker-blobVersionedHash1",
        "opts": {
          "attempts": 3,
          "backoff": {
            "delay": 1000,
            "type": "exponential",
          },
          "jobId": "swarm-worker-blobVersionedHash1",
        },
        "queueName": "swarm-worker",
      },
    ],
    "data": {
      "versionedHash": "blobVersionedHash1",
    },
    "name": "propagateBlob:finalizer-worker-blobVersionedHash1",
    "opts": {
      "attempts": 3,
      "backoff": {
        "delay": 1000,
        "type": "exponential",
      },
      "jobId": "finalizer-worker-blobVersionedHash1",
    },
    "queueName": "finalizer-worker",
  },
  {
    "children": [
      {
        "data": {
          "versionedHash": "blobVersionedHash2",
        },
        "name": "storeBlob:file_system-worker-blobVersionedHash2",
        "opts": {
          "attempts": 3,
          "backoff": {
            "delay": 1000,
            "type": "exponential",
          },
          "jobId": "file_system-worker-blobVersionedHash2",
        },
        "queueName": "file_system-worker",
      },
      {
        "data": {
          "versionedHash": "blobVersionedHash2",
        },
        "name": "storeBlob:google-worker-blobVersionedHash2",
        "opts": {
          "attempts": 3,
          "backoff": {
            "delay": 1000,
            "type": "exponential",
          },
          "jobId": "google-worker-blobVersionedHash2",
        },
        "queueName": "google-worker",
      },
      {
        "data": {
          "versionedHash": "blobVersionedHash2",
        },
        "name": "storeBlob:postgres-worker-blobVersionedHash2",
        "opts": {
          "attempts": 3,
          "backoff": {
            "delay": 1000,
            "type": "exponential",
          },
          "jobId": "postgres-worker-blobVersionedHash2",
        },
        "queueName": "postgres-worker",
      },
      {
        "data": {
          "versionedHash": "blobVersionedHash2",
        },
        "name": "storeBlob:swarm-worker-blobVersionedHash2",
        "opts": {
          "attempts": 3,
          "backoff": {
            "delay": 1000,
            "type": "exponential",
          },
          "jobId": "swarm-worker-blobVersionedHash2",
        },
        "queueName": "swarm-worker",
      },
    ],
    "data": {
      "versionedHash": "blobVersionedHash2",
    },
    "name": "propagateBlob:finalizer-worker-blobVersionedHash2",
    "opts": {
      "attempts": 3,
      "backoff": {
        "delay": 1000,
        "type": "exponential",
      },
      "jobId": "finalizer-worker-blobVersionedHash2",
    },
    "queueName": "finalizer-worker",
  },
]
`;
